src/lock/postgres.rs

```diff
@@ -0,0 +1,182 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+
+use anyhow::{Context, Result};
+use tokio::sync::Mutex;
+use tokio_postgres::{Client, NoTls};
+use tracing::{debug, warn};
+
+use super::DistributedLock;
+
+/// PostgreSQL-based distributed lock using advisory locks.
+///
+/// This lock uses PostgreSQL advisory locks to coordinate access across multiple hosts.
+/// The lock is identified by a 64-bit integer derived from hashing the target path.
+pub struct PostgresLock {
+    client: Arc<Mutex<Client>>,
+    lock_key: i64,
+    target_path: PathBuf,
+    owner: Option<String>,
+    acquired_at: SystemTime,
+}
+
+impl PostgresLock {
+    /// Create a new PostgreSQL lock client.
+    ///
+    /// # Arguments
+    /// * `connection_string` - PostgreSQL connection string (e.g., "host=localhost user=postgres")
+    pub async fn connect(connection_string: &str) -> Result<Arc<Mutex<Client>>> {
+        let (client, connection) = tokio_postgres::connect(connection_string, NoTls)
+            .await
+            .context("failed to connect to PostgreSQL")?;
+
+        // Spawn connection handler
+        tokio::spawn(async move {
+            if let Err(e) = connection.await {
+                warn!(error = %e, "PostgreSQL connection error");
+            }
+        });
+
+        Ok(Arc::new(Mutex::new(client)))
+    }
+
+    /// Convert a path to a lock key by hashing it to a 64-bit integer.
+    fn path_to_lock_key(path: &Path) -> i64 {
+        let mut hasher = DefaultHasher::new();
+        path.hash(&mut hasher);
+        hasher.finish() as i64
+    }
+}
+
+impl DistributedLock for PostgresLock {
+    async fn try_acquire(
+        target: &Path,
+        _ttl: Duration,
+        owner: Option<&str>,
+    ) -> Result<Option<Self>> {
+        // Get PostgreSQL connection string from environment
+        let connection_string = std::env::var("SWANLAKE_LOCK_POSTGRES_CONNECTION")
+            .or_else(|_| std::env::var("PGCONNECTION"))
+            .unwrap_or_else(|_| {
+                // Build connection string from individual env vars
+                let host = std::env::var("PGHOST").unwrap_or_else(|_| "localhost".to_string());
+                let port = std::env::var("PGPORT").unwrap_or_else(|_| "5432".to_string());
+                let user = std::env::var("PGUSER").unwrap_or_else(|_| "postgres".to_string());
+                let dbname =
+                    std::env::var("PGDATABASE").unwrap_or_else(|_| "postgres".to_string());
+
+                let mut conn_str = format!("host={} port={} user={} dbname={}", host, port, user, dbname);
+                if let Ok(password) = std::env::var("PGPASSWORD") {
+                    conn_str.push_str(&format!(" password={}", password));
+                }
+                conn_str
+            });
+
+        let client = Self::connect(&connection_string).await?;
```

Comment:
is it a best practice to connect every time? I thought we connect once and reuse that when acquire lock.